### PR TITLE
Add Zstd/LZ4 support for region files (1.16.4 edition)

### PR DESCRIPTION
**Pull request by egg82**
Zstd using zstd-jni: https://github.com/luben/zstd-jni
LZ4 using lz4-java: https://github.com/lz4/lz4-java

Added new command-line option, --forceWrite, which will forcibly write the chunk/flush to disk regardless of whether or not the chunk has actually changed in a --forceUpgrade.
New compression only happens when a chunk is saved, so running a --forceUpgrade AND --forceWrite may be desirable.

New (hidden) config option for the compression method to use.
There's a new option in PaperConfig (compression level) for Zstd which defaults to 6.
There's a new option in PaperConfig (compression level) for Zlib which defaults to 5.
Added a getter to the API to retrieve the current chunk compression method being used.

Since the patch hooks Mojang's versioning system for new compression methods, it
should be internally consistent, easily-updatable, and reversible with a --forceUpgrade AND a --forceWrite.

This will likely conflict with plugins and programs expecting region files to be compressed using the current standard.

Supercedes #4715

### CODE REVIEW

**Review by Spottedleaf — CHANGES_REQUESTED** [review]
Due to GC concerns with how the library is written, the direct bytebuffer options must be used for ZSTD.

**Comment by egg82** [comment]
@Spottedleaf for posterity, is there a particular issue on the zstd-jni Github that we can link for that?

File: Spigot-Server-Patches/0648-Add-Zstd-LZ4-support-for-region-files.patch [hunk 1]
```diff
@@ -0,0 +1,956 @@
...
++++ b/src/main/java/org/bukkit/craftbukkit/Main.java
+@@ -112,6 +112,7 @@ public class Main {
+                         .describedAs("Yml file");
+ 
+                 acceptsAll(asList("forceUpgrade"), "Whether to force a world upgrade");
++                acceptsAll(asList("forceWrite"), "Whether to force flushing to disk when upgrading"); // Paper - add forceWrite flag
+                 acceptsAll(asList("eraseCache"), "Whether to force cache erase during world upgrade");
+                 acceptsAll(asList("nogui"), "Disables the graphical console");
+ 
+diff --git a/src/main/resources/paper.zstd.dict.0 b/src/main/resources/paper.zstd.dict.0
```
Comment (kashike) [inline Spigot-Server-Patches/0648-Add-Zstd-LZ4-support-for-region-files.patch:350]: Where/how  was this found/generated?

**Review by kashike — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0648-Add-Zstd-LZ4-support-for-region-files.patch (hunk 1 above)
Comment (egg82) [reply Spigot-Server-Patches/0648-Add-Zstd-LZ4-support-for-region-files.patch:350]: Using this project/technique: https://github.com/egg82/CompressionTests#zstd-dictionary
It's not incredibly well-written but it does what it's supposed to do and does it efficiently.

Appending `--dump` will create the raw chunk data for Zstd training, after using Chunky (or similar) to generate chunks.

Comment (egg82) [reply Spigot-Server-Patches/0648-Add-Zstd-LZ4-support-for-region-files.patch:350]: The exact commands I used, for reference:
```Bash
java -jar -Xmx8192m -Xms8192m compression-tests-1.1.2.jar ~/Servers/1.16.4/world/region --dump
find ~/Servers/1.16.4/world/region/rawchunks -type f -print0 | sort -zR | tail -zn +25001 | xargs -0 rm
zstd --train-fastcover ~/Servers/1.16.4/world/region/rawchunks/* -o paper.zstd.dict
```

File: Spigot-Server-Patches/0648-Add-Zstd-LZ4-support-for-region-files.patch [hunk 2]
```diff
@@ -0,0 +1,956 @@
...
+ 
+     public RegionFile(java.nio.file.Path java_nio_file_path, java.nio.file.Path java_nio_file_path1, RegionFileCompression regionfilecompression, boolean flag) throws IOException {
+diff --git a/src/main/java/net/minecraft/server/RegionFileCompression.java b/src/main/java/net/minecraft/server/RegionFileCompression.java
+index 3382d678e68e559b8d3cb9dced4fce24206cd38f..3ac8bf12ff3dc1189f02a6b3b34cb35a68ce2d21 100644
+--- a/src/main/java/net/minecraft/server/RegionFileCompression.java
++++ b/src/main/java/net/minecraft/server/RegionFileCompression.java
+@@ -1,30 +1,122 @@
+ package net.minecraft.server;
+ 
++import com.destroystokyo.paper.PaperConfig; // Paper
```
Comment (MiniDigger) [inline Spigot-Server-Patches/0648-Add-Zstd-LZ4-support-for-region-files.patch:157]: group all these imports in one paper block

Comment (MiniDigger) [reply Spigot-Server-Patches/0648-Add-Zstd-LZ4-support-for-region-files.patch:157]: can also add more imports then, as then the amount doesnt really change anything in regards to maintiance burden

**Comment by MeFisto94** [comment]
In regards of the custom trained dictionary:
How specific is this?
Like will the improvements come with downsides for say citybuild servers having a different block distribution than vanilla?
Or are most gains based on the general region format and thus "content independent"?

**Comment by Titaniumtown** [comment]
I use this patch on my fork and I've had no issues with it whatsoever, it just took a few hours to force upgrade my world. More testing is probably needed though.

**Comment by electronicboy** [comment]
The patch, in general, is fine, afaik the only real remaining issues are tryna deal with the side-effects of ID 10 T

**Comment by Titaniumtown** [comment]
What's `ID 10 T`?

Edit: It's come to my attention I just got played.

**Review by electronicboy — APPROVED** [review]

**Comment by electronicboy** [comment]
Unless there are any objections, I think I'm generally happy with this, at least, as much as I'm gonna be

**Comment by Owen1212055** [comment]
Closing PR, as it seems that this has gone inactive.

If you're still interested in continuing this PR, feel free to rebase and leave a comment and we can reopen it for you. 😄
But since it seems there is a lot of interest in this, if anyone else instead wants to pick up on this additional PRs are welcome!

**Comment by NoahvdAa** [comment]
iirc the merging was held of for two reasons (I believe we discussed this in a VC at some point, if my memory serves me right):
- 1D10T errors; This format is not compatible with vanilla for obvious reasons. Users may not realize this though, and would potentially convert their saves without a proper backup.
- Mojang was (at the time, maybe still is?) thinking about adding different compression formats, and this would possibly clash with that.

**Pull request closed without merging**